Added support for kinesis events

amarrella opened the pull request:
This PR adds support for Kinesis invocations and adds 2 sum types:
InputEvent
https://github.com/EarnestResearch/aws-lambda-haskell-runtime-client/pull/3/files#diff-feaa7dfdb738254f4e8512876fed04e6
OutputEvent
https://github.com/EarnestResearch/aws-lambda-haskell-runtime-client/pull/3/files#diff-79f804137043ed2d7810e0c1a02f58dd

It also modifies Main to prove the approach above.

amarrella commented on app/Main.hs:
```diff
@@ -4,8 +4,7 @@
 
```
Ignore, i know this would replace the api gateway logic

amarrella commented:
@louoso I'm not sure if this is the direction you wanted to take with this project, but I had some free time and I took a stab at this. I'm willing to change anything that you don't see fit :) I'd also like to add `S3` and `Cloudwatch Event` triggers when I have time.

On the output side, it's probably useless to return the same event back for Kinesis & co, so we could have a nullary type or a simple "Success / Failure" kind of message. For now I did it because it was a simple way to debug. Let me know your thoughts on that.

louoso commented:
Very cool. I have some thoughts, but I'm open to input on how best to structure this project.

DISCLAIMER: I only gave this a cursory look , so please correct the following if I've misunderstood.

1) Model types

So it isn't atypical for other lambda libraries to include models for all events. I could certainly see how that is useful, it would be one less thing to write to get started. There is quite a bit of work to do there and some decisions to make like is `ARN` a string or a model itself? I've been deferring on that work assuming that the answers would become more evident as I worked through the various uses for this repo.

It might be worthwhile to generate models like [stratosphere](https://github.com/frontrowed/stratosphere#auto-generation) does.

2) InvocationEvent/OutputEvent

The method to create a `runtimeClient` is only constrained to `FromJSON` and `ToJSON` so you should be able to call `client <- runtimeClient @KinesisDataStreamsEvent @KinesisDataStreamsEvent` without needing the sum types.

3) Echo example

If we add all the possible models, should we also add examples for each model? If so, how should that be organized? Should `stack build` produce an executable for each? Should `make` produce a zip for each? I've deferred on this question too.

amarrella commented:
> So it isn't atypical for other lambda libraries to include models for all events. I could certainly see how that is useful, it would be one less thing to write to get started. There is quite a bit of work to do there and some decisions to make like is `ARN` a string or a model itself? I've been deferring on that work assuming that the answers would become more evident as I worked through the various uses for this repo.

Agreed, maybe we should adopt similar models to what Stratosphere does, if we think they are good practice

> The method to create a `runtimeClient` is only constrained to `FromJSON` and `ToJSON` so you should be able to call `client <- runtimeClient @KinesisDataStreamsEvent @KinesisDataStreamsEvent` without needing the sum types.

Yeah this change made it possible to have a function invoked by Kinesis and API Gateway **at the same time** but maybe it's an anti-pattern in production and we shouldn't encourage it with the sum types. I'm happy to remove them.

> If we add all the possible models, should we also add examples for each model? If so, how should that be organized? Should `stack build` produce an executable for each? Should `make` produce a zip for each? I've deferred on this question too.

If we go for 1 example for each model maybe we can parametrize this instead of building everything all the time (though it would explode a bit the number of executables we create).

If the example is generic enough and it's only matter of changing the input and output type, maybe one example is sufficient, and we can document that others are possible with minimum effort.

amarrella commented:
For now i removed the sum types, and reverted the example back to a simpler one with only api gateway. Note that you don't need the `@APIGatewayInputEvent` if it's in the function signature

louoso commented on src/AWS/Types.hs:
```diff
@@ -0,0 +1,34 @@
+import Data.Text
+import GHC.Generics
+
+data ARN = ARN {
```
This is the `ARN` format for API Gateway, but I don't think it would work other models e.g. kinesis.

amarrella replied:
right! I'll make it more generic

amarrella replied:
Now it's more generic, but i'm unhappy with the decoding, I'll write a parser tomorrow

louoso replied:
I don't think we have to go too deep on the models in this repo. We can customize them as necessary at the call site. In general, I think I'd prefer to close this PR and see if we can code gen all models in a very basic way e.g. `arn :: Text`.

amarrella commented on src/AWS/Types.hs:
```diff
@@ -0,0 +1,34 @@
+
+textToARN :: Text -> Maybe ARN
+textToARN t
+  | (pre : part : svc : reg : accID : res) <- splitOn ":" t
```
Maybe i should get my hands dirty and write a parser for this

louoso replied:
Yeah, Andrey said that too. It would cool, but keep in mind the use case is limited to API Gateway unless you want to make an `ARN` parser that handles all the variations.

louoso commented:
Can you pare this down to just the kinesis events and rebase it to master?

amarrella commented:
Done :)
